src/main/java/org/opensearch/ad/ratelimit/ConcurrentQueue.java

```diff
@@ -0,0 +1,162 @@
+     * @param maintenanceFreqConstant a constant help define the frequency of maintenance.  We cannot do
+     *   the expensive maintenance too often.
+     * @param clientUtil utility with ES client
+     * @param concurrencySetting Max concurrent processing of the queued events
+     * @param executionTtl Max execution time of a single request
+     * @param stateTtl max idle state duration.  Used to clean unused states.
+     * @param nodeStateManager node state accessor
+     */
+    public ConcurrentQueue(
+        String queueName,
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
+        ClusterService clusterService,
+        Random random,
+        ADCircuitBreakerService adCircuitBreakerService,
+        ThreadPool threadPool,
+        Settings settings,
+        float maxQueuedTaskRatio,
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        ClientUtil clientUtil,
+        Setting<Integer> concurrencySetting,
+        Duration executionTtl,
+        Duration stateTtl,
+        NodeStateManager nodeStateManager
+    ) {
+        super(
+            queueName,
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            stateTtl,
+            nodeStateManager
+        );
+
+        this.permits = new Semaphore(concurrencySetting.get(settings));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(concurrencySetting, it -> permits = new Semaphore(it));
+        this.clientUtil = clientUtil;
+
+        this.lastExecuteTime = clock.instant();
+        this.executionTtl = executionTtl;
+    }
+
+    @Override
+    public void maintenance() {
+        super.maintenance();
+
+        if (lastExecuteTime.plus(executionTtl).isBefore(clock.instant())) {
+            LOG.warn("previous execution has been running for too long.  Maybe there are bugs.");
+
+            // release one permit
+            permits.release();
+        }
+    }
+
+    /**
+     * try to execute queued requests if there are concurrency slots and return right away.
+     */
+    @Override
+    protected void triggerProcess() {
+        threadPool.executor(AnomalyDetectorPlugin.AD_THREAD_POOL_NAME).execute(() -> {
+            if (permits.tryAcquire()) {
+                try {
+                    lastExecuteTime = clock.instant();
+                    execute(() -> {
+                        permits.release();
+                        process();
```

Comment:
I am a little confused here. The call to `process()` will call `triggerProcess` again and this will continue until the permit can no longer be acquired?

Reply:
This is a callback after processing one request.  So we keep looking for new requests if there is any after finishing one request.  Otherwise, just release (the 2nd callback).  Added comment to explain.

Reply:
I see that makes sense.